### PR TITLE
a bit messy!

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -98,6 +101,20 @@
 RewardFunc = Union[str, PreTrainedModel, Callable[[list, list], list[float]]]
 
 
+def extract_tool_calls(text: str) -> dict[str, Any]:
+    """
+    Given a list of strings, extract all <tool_call> JSON blocks and return them as a list of dictionaries.
+    """
+    pattern = re.compile(r"<tool_call>\s*(\{.*?\})\s*</tool_call>", re.DOTALL)
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py:109]: There is unfortunately no standardisation for the tool call tags across model families, but Matt is working on extending the chat templates so they can auto-parse tools calls (internal Slack thread): https://huggingface.slack.com/archives/C06JKEMK6BZ/p1757691450090859

Note that `vllm` works around this by providing a dedicated set of parsers that can be set when spinning up the server: https://docs.vllm.ai/en/stable/features/tool_calling.html

I'm not sure we want to go down this route, since it's quite messy in my experience to match the parser to the desired model (e.g. some Qwen models use the `hermes` parser, others not)

So in the meantime, we might want to give uses the ability to provide their own parsing function and default to yours (which is the most common I've seen)

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1413,6 +1434,53 @@ def _generate_and_score_completions(
...
+                new_completion_mask,
+                new_num_items_in_batch,
+                new_sampling_per_token_logps,
+                new_forward_kwargs,
+            ) = self._generate(new_prompts, images)
+            num_tool_ids = new_prompt_mask.sum(-1) - torch.cat(
+                [prompt_mask[needs_tool], completion_mask[needs_tool]], dim=1
+            ).sum(-1)
+            tool_ids = [ids[-num:] for ids, num in zip(new_prompt_ids, num_tool_ids)]
+            tool_mask = [torch.ones_like(ids) for ids in tool_ids]
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py:1464]: Would be cool to have a unit test for this masking so we're confident it is behaving as expected

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (lewtun) [reply trl/trainer/grpo_trainer.py:109]: Perfect timing :) 

https://moon-ci-docs.huggingface.co/docs/transformers/pr_40894/en/chat_response_parsing

Comment (albertvillanova) [reply trl/trainer/grpo_trainer.py:109]: That is the approach we usually followed on smolagents: to provide a sensible default, but allow users to fully customize the function/object instance.

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -98,6 +101,20 @@
 RewardFunc = Union[str, PreTrainedModel, Callable[[list, list], list[float]]]
 
 
+def extract_tool_calls(text: str) -> dict[str, Any]:
```
Comment (albertvillanova) [inline trl/trainer/grpo_trainer.py:105]: What about moving this function to a non-specific trainer module, so it can be used by any trainer in the future?

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (albertvillanova) [inline trl/trainer/grpo_trainer.py:109]: Note that the compile function will be called at each function call. If optimizing performance is necessary in this case, we should set the compilation as a constant at the module level, so it is called only once at import time.

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -214,7 +231,10 @@ def __init__(
         callbacks: Optional[list[TrainerCallback]] = None,
         optimizers: tuple[Optional[torch.optim.Optimizer], Optional[torch.optim.lr_scheduler.LambdaLR]] = (None, None),
         peft_config: Optional["PeftConfig"] = None,
+        tools=None,
     ):
+        self.tools = tools or []
+        self._tool_dict = {name: tool for name, tool in zip([tool.__name__ for tool in self.tools], self.tools)}
```
Comment (albertvillanova) [inline trl/trainer/grpo_trainer.py]: Simpler:
```suggestion
        self._tool_dict = {tool.__name__: tool for tool in self.tools}
```

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -214,7 +231,10 @@ def __init__(
         callbacks: Optional[list[TrainerCallback]] = None,
         optimizers: tuple[Optional[torch.optim.Optimizer], Optional[torch.optim.lr_scheduler.LambdaLR]] = (None, None),
         peft_config: Optional["PeftConfig"] = None,
+        tools=None,
```
Comment (albertvillanova) [inline trl/trainer/grpo_trainer.py:249]: You will need to add the `tools` param to the trainer docstring. And give a type hint.

File: trl/trainer/grpo_trainer.py [hunk 6]
```diff
@@ -98,6 +101,20 @@
...
 
+def extract_tool_calls(text: str) -> dict[str, Any]:
+    """
+    Given a list of strings, extract all <tool_call> JSON blocks and return them as a list of dictionaries.
+    """
+    pattern = re.compile(r"<tool_call>\s*(\{.*?\})\s*</tool_call>", re.DOTALL)
+
+    for match in pattern.findall(text):
+        try:
+            return json.loads(match)
```
Comment (albertvillanova) [inline trl/trainer/grpo_trainer.py:113]: You only return the first match?

File: trl/trainer/grpo_trainer.py [hunk 7]
```diff
@@ -1405,6 +1426,53 @@ def _generate_and_score_completions(
             sampling_per_token_logps,
             forward_kwargs,
         ) = self._generate(prompts, images)
+        completions = self.processing_class.batch_decode(completion_ids, skip_special_tokens=True)
+        tool_calls = [extract_tool_calls(completion) for completion in completions]
+        tool_results = [self._tool_dict[tc["name"]](**tc["arguments"]) if tc else None for tc in tool_calls]
+        tool_messages = [
+            [{"role": "tool", "name": tc["name"], "content": str(tr)}] if tc else None
+            for tc, tr in zip(tool_calls, tool_results)
+        ]
```
Comment (albertvillanova) [inline trl/trainer/grpo_trainer.py:1445]: Not sure of this handles potential multiple tool calls in a single completion...

Comment (albertvillanova) [inline trl/trainer/grpo_trainer.py:1445]: Before the messages with the tool results (`"role": "tool"`), shouldn't we prepend the messages with the tool calls themselves (`"role": "assistant", "tool_calls":...`)? Not sure of this though... A real question! 😅

**Comment by August-murr** [comment]
I suggest that we remove any references to including tools as a parameter in the `GRPOTrainer`:

```
trainer = GRPOTrainer(…,
tools = [tool_1, tool_2]
)
```

While I am not necessarily against chat templates or tool parsing implementations, I believe the attribute `tools` should be eliminated.

To maximize scalability, a better approach would be to create an all-in-one tools sandbox. This means that the only tool the training script interacts with would be the sandboxed code executors, with all the necessary tools defined within it, like using a Docker image that contains the dependencies of the tools to initialize the sandbox 

Therefore, I propose that we only add the parameter `Environment`, which would encompass the code executor along with all initialized tools.

**The environment needs to be responsible for handling tool usage and execution.**

Then you could create your own built-in environments using your preferred parsing and chat templates.

**Comment by qgallouedec** [comment]
Yes, I have seen some works that suggest that when scaling up, multiplying tools works less well than a smaller, more generic set of tools, such as the one you describe. However, it seems to me that the approach proposed here is in fact compatible:

```python
trainer = GRPOTrainer(…,
    tools = [my_big_containerized_all_in_one_tool],
)
```

in the end it's up to the user decide how to design it, most important here is to allow for this flexibility

**Comment by August-murr** [comment]
but the name `tools` as a param in `GRPOTrainer` is misleading.
What I suggest is more like
```py
MyEnv = DefaultEnv(code_executer=my_big_containerized_all_in_one_tool)

trainer = GRPOTrainer(…,
Environment=MyEnv
)

```
The user can then customize tool use in their own environment.

**Comment by qgallouedec** [comment]
the name tool is actually quite standard now:

https://huggingface.co/blog/unified-tool-use

https://huggingface.co/docs/transformers/main/chat_extras

https://huggingface.co/docs/transformers/main/en/internal/tokenization_utils#transformers.PreTrainedTokenizerBase.apply_chat_template.tools

File: trl/trainer/grpo_trainer.py (hunk 4 above)
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: it must have been very late 😂 😂

**Comment by qgallouedec** [comment]
closed in favor of #4300
